### PR TITLE
New implementation of metadata fixer

### PR DESCRIPTION
**Pull request by ZhenyaM**

### CODE REVIEW

File: metadata-fixer/src/main/java/org/verapdf/metadata/fixer/MetadataFixer.java [hunk 1]
```diff
@@ -221,200 +160,141 @@ public FixReport fixDocument(OutputStream output) throws IOException, URISyntaxE
...
-		if (parser == null) {
-			parser = XMLProcessedObjectsParser.getInstance();
+	private static FixReport getValidationStatus(FixerConfig config) {
+		Result result = config.getValidationResult();
+		if (result != null) {
+			List<Rule> rules = result.getDetails().getRules();
+			ValidationProfile profile = config.getValidationProfile();
+			if (profile != null) {
+				try {
+					return ProcessedObjectsInspector.validationStatus(rules, profile, config.getParser());
```
Comment (timurkamalov) [inline metadata-fixer/src/main/java/org/verapdf/metadata/fixer/MetadataFixer.java]: you know

File: metadata-fixer/src/main/java/org/verapdf/metadata/fixer/MetadataFixer.java [hunk 2]
```diff
@@ -221,200 +160,141 @@ public FixReport fixDocument(OutputStream output) throws IOException, URISyntaxE
...
 
-	private void fixProperty(FixReport entity, XMPSchema schema, String metaValue,
-							 String infoValue, String attribute) {
+	private static void fixProperty(FixReport entity, BasicSchema schema, InfoDictionary info, String metaValue,
+									String infoValue, String attribute) {
 		String key = attributes.get(attribute);
 		if (metaValue == null && infoValue != null) {
 			doSaveAction(schema, attribute, infoValue);
-			entity.addFix("Add " + key + " to metadata from info dictionary");
+			entity.addFix("Add '" + key + "' to metadata from info dictionary");
```
Comment (timurkamalov) [inline metadata-fixer/src/main/java/org/verapdf/metadata/fixer/MetadataFixer.java]: added

File: metadata-fixer/src/main/java/org/verapdf/metadata/fixer/impl/pb/FixerConfigImpl.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+	@Override
+	public ValidationProfile getValidationProfile() {
+		return validationResult == null ? null :
+				validationResult.getProfile().getValidationProfile();
+	}
+
+	@Override
+	public Metadata getMetadata() {
+		return document.getMetadata();
```
Comment (timurkamalov) [inline metadata-fixer/src/main/java/org/verapdf/metadata/fixer/impl/pb/FixerConfigImpl.java]: this

File: metadata-fixer/src/main/java/org/verapdf/metadata/fixer/impl/pb/model/MetadataImpl.java [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+/**
+ * @author Evgeniy Muravitskiy
+ */
+public class MetadataImpl implements Metadata {
+
+	private final XMPMetadata metadata;
+	private final COSStream stream;
+
+	public MetadataImpl(XMPMetadata metadata, COSStream stream) {
+		this.metadata = metadata;
```
Comment (timurkamalov) [inline metadata-fixer/src/main/java/org/verapdf/metadata/fixer/impl/pb/model/MetadataImpl.java]: check for null

File: metadata-fixer/src/main/java/org/verapdf/metadata/fixer/impl/pb/model/PDFDocumentImpl.java [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+public class PDFDocumentImpl implements PDFDocument {
+
+	private static final Logger LOGGER = Logger.getLogger(PDFDocumentImpl.class);
+
+	private final PDDocument document;
+	private MetadataImpl metadata;
+	private InfoDictionaryImpl info;
+
+	public PDFDocumentImpl(PDDocument document) {
+		this.document = document;
```
Comment (timurkamalov) [inline metadata-fixer/src/main/java/org/verapdf/metadata/fixer/impl/pb/model/PDFDocumentImpl.java]: check for null

File: metadata-fixer/src/main/java/org/verapdf/metadata/fixer/impl/pb/schemas/AdobePDFSchemaImpl.java [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+/**
+ * @author Evgeniy Muravitskiy
+ */
+public class AdobePDFSchemaImpl implements AdobePDF {
+
+	private final AdobePDFSchema schema;
+	private final Metadata metadata;
+
+	public AdobePDFSchemaImpl(AdobePDFSchema schema, Metadata metadata) {
+		this.schema = schema;
```
Comment (timurkamalov) [inline metadata-fixer/src/main/java/org/verapdf/metadata/fixer/impl/pb/schemas/AdobePDFSchemaImpl.java]: check

File: metadata-fixer/src/main/java/org/verapdf/metadata/fixer/impl/pb/schemas/DublinCoreSchemaImpl.java [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+ */
+public class DublinCoreSchemaImpl implements DublinCore {
+
+	private static final String AUTHOR = "creator";
+
+	private final DublinCoreSchema schema;
+	private final Metadata metadata;
+
+	public DublinCoreSchemaImpl(DublinCoreSchema schema, Metadata metadata) {
+		this.schema = schema;
```
Comment (timurkamalov) [inline metadata-fixer/src/main/java/org/verapdf/metadata/fixer/impl/pb/schemas/DublinCoreSchemaImpl.java]: check

File: metadata-fixer/src/main/java/org/verapdf/metadata/fixer/impl/pb/schemas/DublinCoreSchemaImpl.java [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+		List<String> creators = this.schema.getCreators();
+		return creators != null && creators.size() > 0 ? creators.get(0) : null;
+	}
+
+	@Override
+	public void setAuthor(String creator) {
+		ArrayProperty seq = (ArrayProperty) schema.getAbstractProperty(AUTHOR);
+		ArrayProperty newSeq = schema.createArrayProperty(AUTHOR, Cardinality.Seq);
+		TextType li = schema.createTextType(XmpConstants.LIST_NAME, creator);
+		newSeq.getContainer().addProperty(li);
```
Comment (timurkamalov) [inline metadata-fixer/src/main/java/org/verapdf/metadata/fixer/impl/pb/schemas/DublinCoreSchemaImpl.java]: null?

File: metadata-fixer/src/main/java/org/verapdf/metadata/fixer/utils/FileGenerator.java [hunk 9]
```diff
@@ -72,31 +56,31 @@ public static File createOutputFile(String sourcePath, String prefix) {
...
 	 * @param fileName   name of the result file
 	 * @param prefix     prefix for the result file name
 	 * @return non-existing file which stored near source file
-	 * @throws IllegalArgumentException if {@code folderPath} or
+	 * @throws IllegalArgumentException if {@code folderFile} or
 	 * {@code fileName}, or {@code prefix} are {@code null}
 	 */
-	public static File createOutputFile(String folderPath, String fileName, String prefix) {
-		if (folderPath == null) {
+	public static File createOutputFile(File folderFile, String fileName, String prefix) throws IOException {
```
Comment (timurkamalov) [inline metadata-fixer/src/main/java/org/verapdf/metadata/fixer/utils/FileGenerator.java]: remove

File: metadata-fixer/src/main/java/org/verapdf/metadata/fixer/impl/pb/schemas/DublinCoreSchemaImpl.java (hunk 8 above)
Comment (ZhenyaM) [inline metadata-fixer/src/main/java/org/verapdf/metadata/fixer/impl/pb/schemas/DublinCoreSchemaImpl.java]: Container always not null:
public AbstractComplexProperty(XMPMetadata metadata, String propertyName)
    {
        super(metadata, propertyName);
        container = new ComplexPropertyContainer();
        this.namespaceToPrefix = new HashMap<String, String>();
    }

**Pull request merged**